// @alwatr/type/customer-order-management.js
// -- Const value --
export const ladingTypeCS = ['hand', 'pallet'];
export const carTypeCS = [
    'trailer_truck',
    'camion_dual',
    'camion_solo',
    'camion_911',
    'camion_800',
    'camion_600',
    'camion_mini',
    'nissan',
];
export const timePeriodCS = ['auto', '3_4w', '2_3w', '1_2w'];
export const discountTypeCS = ['number', 'percent'];
export const orderStatusCS = [
    'draft',
    'registered',
    'processing',
    'payment_pending',
    'preparing',
    'shipping',
    'delayed',
    'on_hold',
    'canceled',
    'refunded',
];
export const userPermissionsCS = ['user/patch', 'price/patch', 'product/patch', 'user-list-inc-order/read'];
// -- Schema --
export const orderInfoSchema = {
    id: String,
    status: String,
    itemList: [
        {
            productId: String,
            marketPrice: Number,
            agencyPrice: Number,
            qty: Number,
        },
    ],
    shippingInfo: {
        recipientName: String,
        recipientNationalCode: String,
        address: String,
        carType: String,
        ladingType: String,
        timePeriod: String,
        // description: String,
    },
    // discount: Number,
    // discountType: String,
    subTotalMarket: Number,
    subTotalAgency: Number,
    ladingFee: Number,
    palletCost: Number,
    shippingFee: Number,
    totalShippingFee: Number,
};
export const orderShippingInfoSchema = orderInfoSchema.shippingInfo;
